### DISCUSSION TITLE
Places: Opening the photo viewer requires too much memory when there are many results

### DISCUSSION THREAD

**Discussion by vb0** [photoprism/photoprism#1707] [Q&A] [answered]
Click on places, wait for map to load, zoom in, zoom in, make sure there aren't loads of pictures stacked (like I see a circle with 5 inside, zoom in), I see round thumbnails, all fine. Debug logs on the server with all GETs/database queries, etc. All of the above is fine and I guess that should be the more intensive part both on the server and browser. 

Then click on one picture - nothing happens except the browser eating more and more RAM until it runs out. Server log is silent, not much network activity (the server is fine and can serve other clients, only this tab is gone).  

I took it as a regular "more RAM needed" until I went to 32GB and still not enough. Windows + latest stable Chrome and Firefox were tested.  

This might be a regression, it used to work (not flawless but it was working some months ago).

Large library - Search shows 477k, Places shows 1400 places.
Using latest docker linux version.

**Comment by lastzero** [comment, marked as answer]
Maybe you had less images months ago? I can imagine that creating the data structure holding 477k images is too much for the current photo viewer. It's can't dynamically load more pictures, so the list is generated upfront (pictures are NOT all loaded upfront). Can you try what happens if you search for a keyword or place? This reduces the number of pictures.

**Reply by vb0** [reply]
Yes, this is, thank you for the quick answer. It looks like "zooming in" actually doesn't limit the pictures that are loaded in the list for the photo viewer, at least for the quick example I can try with the following workflow:  

* filter for placeX where I get just a few results (in this case I have 3 groups of 1+11+4 pics)  
* no matter how I zoom in and how much is on the current view of the map the viewer cycles through the same (all) 16 pics (but at least it works)  

I'm not sure this is the intended behavior, even if there are no performance concerns if the user carefully zooms in to I don't know, Tower Bridge (London) and clicks on one of the pics there he shouldn't then cycle through all his pics from all continents, even if he has "just" 5k pics it's still too bad for him.

**Reply by lastzero** [reply]
Yes, that's the way it works as the map items are loaded once and rendered dynamically. The photo viewer is not part of the map, so its contents are not dynamic.

**Reply by lastzero** [reply]
Tested with 100k images on OSX with just 16 GB of RAM and Google Chrome. Opening the viewer is a bit slow, but definitely not running out of memory.

**Reply by lastzero** [reply]
Obviously we will improve this and a new image viewer is already on our roadmap. It's just that we can't work on everything at the same time.

**Reply by lastzero** [reply]
To get a dynamic picture list, you'd need to get the visible object list from the memory of the map viewer as it depends on the coordinates, viewport size, view angle, and zoom level.

**Reply by lastzero** [reply]
You are also welcome to send a PR with optimized JS code as a temporary workaround until a new viewer is available!
